lib/web_of_science/harvester.rb

```diff
@@ -52,75 +58,72 @@ def process_pmids(author, pmids)
+      process_records author, queries.retrieve_by_pmid(pmids)
     end
 
     # Harvest WOS-UID publications for an author
     # @param author [Author]
-    # @param uids [Array<String>] WOS-UID or WOS-ItemId values (not URIs)
+    # @param uids [Array<String>] WOS-UID values (not URIs)
     # @return [Array<String>] WosUIDs that create Publications
     def process_uids(author, uids)
       raise(ArgumentError, 'author must be an Author') unless author.is_a? Author
       raise(ArgumentError, 'uids must be Enumerable') unless uids.is_a? Enumerable
-      uids.reject! { |uid| publication_identifier?('WosItemId', wos_item(uid)) }
+      uids -= author_contributions(author, uids)
       return [] if uids.empty?
-      process_records author, wos_queries.retrieve_by_id(uids)
+      process_records author, queries.retrieve_by_id(uids)
     end
 
     private
 
+      delegate :logger, to: :WebOfScience
+      delegate :queries, to: :WebOfScience
+
       # Process records retrieved by any means
       # @param author [Author]
       # @param records [WebOfScience::Records]
       # @return [Array<String>] WosUIDs that create Publications
       def process_records(author, records)
+        return [] if records.empty?
         processor = WebOfScience::ProcessRecords.new(author, records)
         processor.execute
       end
 
       # ----
-      # Retrieve WOS Records for Author
+      # Find or create contributions for WOS Records by Author
 
+      # Find any matching contributions by author and WOS-UID - when a publication exists but
+      # the author is not yet a contributor to it, add the contribution.
       # @param author [Author]
-      # @return [WebOfScience::Records]
-      def records_for_author(author)
-        # TODO: iterate on author identities also, or leave that to the consumer of this class?
-        names = author_name(author).text_search_query
-        institution = author_institution(author).normalize_name
-        user_query = "AU=(#{names}) AND AD=(#{institution})"
-        wos_queries.search(user_query)
+      # @param uids [Array<String>]
+      # @return [Array<String>]
+      def author_contributions(author, uids)
+        # TODO: consider any more efficient queries to find/create contributions
+        Publication.where(wos_uid: uids).map do |pub|
+          contrib = pub.contributions.find_or_initialize_by(author_id: author.id)
+          author_contribution_saved?(author, contrib) ? pub.wos_uid : nil
+        end.compact
       end
 
+      # Save a new contribution to a WOS-UID publication for author.
       # @param author [Author]
-      # @return [Agent::AuthorName]
-      def author_name(author)
-        Agent::AuthorName.new(
-          author.last_name,
-          author.first_name,
-          Settings.HARVESTER.USE_MIDDLE_NAME ? author.middle_name : ''
+      # @param contrib [Contribution]
+      # @return [Array<String>]
+      def author_contribution_saved?(author, contrib)
```

Comment:
This is a hazardous conflation of a boolean-like interrogative method and a an actual edit/`.save!`.  The documented return value appears to be incorrect.

Reply:
OK, changed.